router client, server client written and tested. integration testing run

dostapenko opened the pull request.

Commit:
"server steps deleted and all scenarios are represented in rspec"

Commit:
"anomalies steps deleted. All the tests are represented in rspec tests"

Commit:
"several TODOs accomplished"

Commit:
"deletion of old code which is no longer used"

Commit:
"deletion of old code which is no longer used in server"

Commit:
"configuration spec reworked and sandbox changed to IPN_generator"

Commit:
"router client written and router rspec tests passing"

Commit:
"load config rspec spelling corrections"

Commit:
"forward ipn steps deleted"

Commit:
"test cleanup. All tests running again"

Commit:
"server client spec written"

Commit:
server client rspec tests fleshed out and passing

Commit:
removal of server code which now lives in server client

Commit:
"deletions of old server-computer handshake interactions"

Commit:
Merge branch 'master' of https://github.com/hedgeyedev/paypal-ipn-forwarder into deletion_of_verified_messages_for_dev_comp

Commit:
"fleshing of remaining rspec tests"

Commit:
Merge branch 'master' of https://github.com/hedgeyedev/paypal-ipn-forwarder into readme

Commit:
"sequence diagrams added"

Commit:
readme edited

Commit:
merge conflicts resolved

Commit:
config.ru modification for integration test

Commit:
rspecs passing

Commit:
Merge branch 'rspec_fix' into server_client_fleshing

Commit:
test mode turning on

Commit:
localhost integration test passed

Commit:
localhost integration test passed with app running and rspec tests pass

Commit:
computer alerts written, with rspecs and cukes deleted

Commit:
start_paypal written and working

Commit:
'minor changes for readability

dostapenko commented on lib/ipn_generator.rb:
```diff
@@ -15,20 +16,23 @@ def ipn
+  end
+
+
+  def send_via_http(url)
```
used to send an IPN to start testing for config.ru

dostapenko commented on lib/router.rb:
```diff
@@ -9,48 +8,30 @@ class Router
-
+   #TODO: rename dev_computer to router Client
   def initialize(target, test=nil)
-    @development_computer  = target
```
a lot of functionality now in router_client. Router_client handles Rest_client interactions

dostapenko commented on lib/server.rb:
```diff
@@ -24,31 +22,11 @@ def initialize(test=nil)
-  def receive_poll_from_computer(paypal_id)
-  end
-
-  def send_ipn_if_present(paypal_id)
```
functionality in server_client. Server_client handles interactions between confi.ru and server

dostapenko commented on lib/server.rb:
```diff
@@ -24,31 +22,11 @@ def initialize(test=nil)
     params['receiver_id'].first
   end
 
-  def paypal_email_id_map(paypal_email)
```
no longer used

dostapenko commented on lib/server.rb:
```diff
@@ -102,35 +81,26 @@ def send_conflict_email(paypal_id, email)
      @email_map[id] = email
   end
 
-  def computer_online?(id)
```
method moved to line 42, not removed

dostapenko commented on lib/server.rb:
```diff
@@ -159,31 +130,21 @@ def ipn_present?(paypal_id)
     queue_size(paypal_id) >= 1
   end
 
-  def receive_ipn_response(ipn_response)
```
no longer using IPN responses from the developer computer. Those should be muted in the app

dostapenko commented on lib/server.rb:
```diff
@@ -159,31 +130,21 @@ def ipn_present?(paypal_id)
 
-  def respond_to_computer_poll(paypal_id, now=Time.now)
-    #a new instance of poll checker needs to be created in case poll is before test mode is turned on and the sandbox is not registered beforhand
+  def record_computer_poll(paypal_id)
```
following 2 poll methods are called from server_client

dostapenko commented on spec/development_computer_spec.rb:
```diff
@@ -1,24 +0,0 @@
-require 'rspec'
```
development_computer was deleted. router_client is the closest to it's equivalent

dostapenko commented on spec/router_spec.rb:
```diff
@@ -3,50 +3,50 @@
 
   context 'interactions with server' do
 
+    it 'can not reach the server when starting test mode' do
```
will change to 'developer notified that it'...

dostapenko commented on spec/router_spec.rb:
```diff
@@ -56,58 +56,26 @@ def expected_rest_client_message(mode)
-      @router.forward_ipn(ipn)
-
-    end
+      @router_client.stub!(:send_ipn).with(ipn).and_return(ipn_response)
```
response is a mistake. Will get rid of it

Commit:
changes based on PR mistakes found

Commit:
minor change

Commit:
'changes to ensure that nil paramaters arent passed in'

Commit:
most rspec tests passing

dostapenko commented on bin/router:
```diff
@@ -1,18 +0,0 @@
-#!/usr/bin/env ruby
```
changed implementation so that this is no longer used. Now, start_paypal is used to start the gem on the dev computer and the forked process is just the polling, which is all done in the start_paypal script

dostapenko commented on config.ru:
```diff
@@ -1,108 +1,71 @@
 
 class ServerRack < Sinatra::Base
-
-  def initialize(server = Server.new)
```
initialize wasnt working during testing. configure was the way to go

dostapenko replied:
I could not get the methods to work, regardless of what I did so I got rid of them and didnt use them. The best working solution i got to was using self.method_name but then they couldnt be called. Any idea if there is a way to make this work correctly? I couldnt find any solutions

dostapenko commented on config.ru:
```diff
@@ -1,108 +1,71 @@
-  def launch_ipn
-    cause_IPN_post_staement_against "localhost:8810/ipn/payemtns"
+  configure do
+    TEST_MODE_ON = true
```
I will take this out for later version or make it false. I was using it to load in the test yaml file rather than the real one

dostapenko commented on config.ru:
```diff
@@ -1,108 +1,71 @@
   end
 
   get '/invoke_ipn' do
-    launch_ipn
```
used for testing. acted as sandbox, sending the IPN to the "server"

dostapenko commented on features/configuration.feature:
```diff
@@ -4,15 +4,3 @@ Feature: Match PayPal sandboxes with developer computers
   So I can use the sandbox confident that it is not receiving other developers' requests.
   #TODO: Need glossary for 'computer', 'server'
 
```
features are deleted part of cuke cleaning/removal

dostapenko commented on lib/router.rb:
```diff
@@ -9,48 +8,29 @@ class Router
-  end
-
-
-  def forward_ipn(ipn)
```
dealing with dev-computer-issued verified messages. No longer using this implementation

dostapenko commented on lib/router.rb:
```diff
@@ -9,48 +8,29 @@ class Router
-
-  def send_verified
-    @development_computer.send_verified
+    @router_client  = target
```
router client takes care of all the rest-client messages sent from router

dostapenko commented on lib/server.rb:
```diff
@@ -61,35 +39,36 @@ def ipn_response(ipn)
     response
   end
 
-  def computer_testing(params)
```
computer-testing as a method is in server_client. That method makes calls to the new ones below

dostapenko commented on lib/server.rb:
```diff
@@ -159,31 +130,26 @@ def ipn_present?(paypal_id)
+    @poll_checker_instance[paypal_id].unexpected_poll_time(paypal_id, time)
   end
 
-  def respond_to_computer_poll(paypal_id, now=Time.now)
```
moved to server_client which makes calls to these new methods

dostapenko commented on lib/server.rb:
```diff
@@ -159,31 +130,26 @@ def ipn_present?(paypal_id)
-    else
-      send_ipn_if_present(paypal_id)
-    end
+  def poll_with_incomplete_info(email, test_mode, id)
```
called in case test mode turned on/off with one of the parameters missing/blank

dostapenko commented on lib/server_poll_checker.rb:
```diff
@@ -50,13 +56,27 @@ def check_testing_polls_occurring(paypal_id, time=@content.poll_checking_interva
    end
   end
 
+  def email_developer_incomplete_request(email, test_mode, id, time=Time.now)
```
this method is to be used when not all the parameters in a test_mode_on post are present. So if the sandbox id or mode is missing, itll send it out

dostapenko commented on spec/server_client_spec.rb:
```diff
@@ -1,10 +1,68 @@
 
-  it 'should turn off testing'
+
+    #TODO: fix tests so the pass with CGI
```
CGI wasnt used on paramaters in the post requests. Changed in implementation but not in the tests yet

dostapenko commented on spec/server_poll_checker_spec.rb:
```diff
@@ -48,5 +57,11 @@
   end
+
+  it 'should not send an email if last incomplete information poll was less than hour ago' do
+      Pony.should_receive(:mail).with(any_args)
```
2 incomplete test-mode-turned-on messages. But only 1 email sent out

dostapenko commented on spec/server_spec.rb:
```diff
@@ -102,37 +72,55 @@
 
   end
 
-  it 'receives a "test mode on" message for a paypal sandbox which is already being used for IPN testing' do
```
test now in server_client spec. not removed completely

Commit:
changes based on comments on PR

Commit:
serverclient rspec tests all passing

Commit:
all rspec tests passing

oldfartdeveloper commented on config.ru:
```diff
@@ -1,108 +1,73 @@
-    @server.respond_to_computer_poll(params)
+  get '/computer_poll' do
+    params = request['sandbox_id']
+    @@server_client.respond_to_computer_poll(params) if !params.nil? && param != ''
```
Suggest:  (DeMorgan's theorem transformation is easier for me to read at least)

```
 @@server_client.respond_to_computer_poll(params) unless params.nil? || params.length == 0
```

oldfartdeveloper commented on config.ru:
```diff
@@ -1,108 +1,73 @@
-    unless ipn == 'VERIFIED' || ipn == 'INVALID'
-      @server.receive_ipn(ipn)
-      response = @server.ipn_response(ipn)
+    if ipn != 'VERIFIED' && ipn != 'INVALID' && ipn != '' && !ipn.nil?
```
Again, DeMorgan is your friend.  This time regular expressions make this much easier to read (and write) as well:

```
 unless ipn.nil? || ipn =~ /(|VERIFIED|INVALID)/
```

Note how `/(|VERIFIED|INVALID)/` substitutes nicely for `ipn != 'VERIFIED' && ipn != 'INVALID' && ipn != ''`.

oldfartdeveloper commented on config.ru:
```diff
@@ -1,108 +1,73 @@
   end
 
   # Pretend to be the PayPal sandbox you're sending the response back to
   post '/fake_payal' do
```
`fake_payal` ==> `fake_paypal`

oldfartdeveloper commented on config.ru:
```diff
@@ -1,108 +1,73 @@
-ross=19.95&shipping=0.00
-EOF
-    RestClient.post url, sample_ipn
+    "Hello World"
```
Why are you not doing a 404 here?  Haven't gotten to it yet?

oldfartdeveloper commented on config.ru:
```diff
@@ -1,108 +1,73 @@
-    ipn = "_notify-validate&" + ipn
-    RestClient.post url, ipn
+  get '/turn_testing_on' do
+    @@router.set_test_mode('on', 'bob@example.com' , 'my_sandbox_id')
```
This seems like obsolete code.  How do you propose to identify who is making this request to turn testing on?  Contact me if you need ideas.

dostapenko replied:
This isnt obsolete. This paypal_sandbox id is used to identify the sandbox being tested. The person doesnt interest us(except in error cases, in which case the email is used as an identifier and is used to contact the person)

Commit:
all rspec tests passing

Commit:
PR comment corrections

Commit:
all tests passing and process no longer forks during rspec testing

Commit:
TODO finished and all rspec tests passing

Commit:
paypal_testing_off written

Commit:
Merge branch 'condensed_readme' into server_client_fleshing

Commit:
minor changes to last cuke tests

Commit:
merge from todo_fixes

Commit:
router feature deleted

Commit:
cukes all removed

Commit:
all rspec tests passing

oldfartdeveloper committed:
[#51739833] Trying relative addressing for URL

oldfartdeveloper committed:
[#51739833] Relative addressing for URL works so I converted all of the image links to relative

oldfartdeveloper committed:
[#51739833] Relative addressing for URL works only if you append '?raw-true' to end of relative URL.  So testing on one.

oldfartdeveloper committed:
[#51739833] Relative addressing; still testing '?raw=true'

oldfartdeveloper committed:
[#51739833] Trying a different GitHub instruction: don't preface relative URI w/ slash

oldfartdeveloper committed:
[#51739833] Trying a different GitHub instruction: don't preface relative URI w/ slash and add '?raw=true'

oldfartdeveloper committed:
[#51739833] So it's an incomplete fix in that you have to relative through the branch.  So trying that.

oldfartdeveloper committed:
[#51739833] So it's an incomplete fix in that you have to relative through the branch.  Make a mistake in the path which is now corrected

oldfartdeveloper committed:
[#51739833] So it's an incomplete fix in that you have to relative through the branch.  Correct another path mistake

oldfartdeveloper committed:
[#51739833] I keep trying

oldfartdeveloper committed:
Stick w/ the GitHub way of doing things and try changing the header in the svg file

Per http://stackoverflow.com/questions/13808020/include-an-svg-hosted-on-github-in-markdown?rq=1

oldfartdeveloper committed:
Remove XML title tag

Per http://stackoverflow.com/questions/13808020/include-an-svg-hosted-on-github-in-markdown?rq=1

oldfartdeveloper committed:
Testing svg file over on gh-pages

Per http://stackoverflow.com/questions/13808020/include-an-svg-hosted-on-github-in-markdown?rq=1

oldfartdeveloper committed:
Now referencing gh-pages/_doc to avoid involving Jekyll

oldfartdeveloper committed:
Trying on the 2nd svg file

oldfartdeveloper committed:
Keep playing

oldfartdeveloper committed:
Added gem github-pages

oldfartdeveloper committed:
Merge branch 'server_client_fleshing' of github.com:oldfartdeveloper/paypal-ipn-forwarder into server_client_fleshing

oldfartdeveloper committed:
Keep trying

oldfartdeveloper committed:
Correct bad naming and try again

oldfartdeveloper committed:
Testing pointing to gh-pages subdirectory

oldfartdeveloper committed:
See if there's a change; problem is in this file

I think the underscore in front of _doc is the problem

oldfartdeveloper committed:
Subdirectories for .svg files a problem?

oldfartdeveloper committed:
[#51738699] Showed how 'glossary' could look; other formatting cleanup

oldfartdeveloper committed:
[#51738699] Merged w/ origin

Commit:
Merge pull request #4 from oldfartdeveloper/server_client_fleshing

Update README.md with several formatting features

Commit:
seq diagram's urls changed to work

Commit:
glossary expanded and components put under their "over-lords"

oldfartdeveloper commented:
Unfortunately, I've left this open too long, so I'm going to accept it as is, especially since we're testing off the branch anyways.

:thumbsup:
